### DISCUSSION TITLE
Add high pwm freq support

### DISCUSSION THREAD

**Discussion by Ncerzzk** [ExpressLRS/ExpressLRS#1810] [Ideas]
As I post before, I tried making a ELRS RX module which could drive brush motor directly(with MOS FET), which could greatly help to make the micro plane.

The hardware has finished here: https://github.com/AutopilotPi/Mini_ELRS ,
and the software changes have been made here:https://github.com/Ncerzzk/ExpressLRS

main changes:
- add high pwm freq support(for now, becasue we set resolution_bits of ledc to fix 16, so the highest freq is 1.2k)
- add duty mode(so that we could map the channels output(pulse width:1000us - 2000us) to 0-100%)
- extend narrow to 3 bits, and add normal mode(0), half mode(1), duty mode(2)  inside it. (may be need a more proper name to replace narrow?)

while the SW changes are still WIP, I would do more test later.

![20220901235840](https://user-images.githubusercontent.com/9284611/187959912-d712891d-778b-4b42-bc26-ba42fe778d6a.png)

**Comment by pkendall64** [comment]
Your changes look quite promising. I've added a couple of minor comments to github for you. I'll take a closer look when you get it ready for PR.

**Reply by Ncerzzk** [reply]
Hi, Paul. I have almost finished the main work, I test it work normaly.
The main recent change is:

add `getLedcChn` to map pwm channels to ledc channel. 
according to the implement before(ledc channel  0 map to pwm channel 0, ledc ch1 to pwm ch1...), the freq of ch0 would be determined by ch1, freq of ch2 would be determined by ch3, because they use the same ledc timer.
In my change, I do this map:
pwmchn     ledcchn
0    0
1    2
2    4
3    6
4    1
5    3
6    5
7    7

so we can get 4 different freq at ch0-ch3, I think these 4 channels are more important?

**Reply by Ncerzzk** [reply]
there are still a little code to change:
- change variable name of `narrow mode`, may be `Pulse Width Mode`?
- change some functions define using `bool narrow`

how about your opinion? Thank you.

**Reply by pkendall64** [reply]
I see whats going on with the timers now. So, there are 4 timers in total (0-3). I think it would be better is we had a timer manager so we could allocate timers to a frequency and all channels using that frequency could use that timer.
Also, bear in mind that timer 3 may be used for the RGB LED. So really only 3 timers are available for different frequencies.
I think it more likely that a user would have the same frequency servo for most surfaces so sharing the timer would be better.

**Reply by Ncerzzk** [reply]
Thank you, but I didn't find where RGB LED is using timer3, can you show me?

**Reply by Ncerzzk** [reply]
Hi, @pkendall64 . 
I add a new func `allocateLedcChn`  to allocate the timers. (https://github.com/Ncerzzk/ExpressLRS/commit/c6dbf5fe4e1349c0b1b9c0cc31c745ed82f38214)

I have done some test.  In ESP32, there are 8 timer(4 high speed, 4 lower), 16 ledc channels, and all the 8 timer could use 80Mhz as the clk source. In the pwm scenes, it seems no the difference between high speed timer and lower. So I allocate low speed timer first, to leave space for people may use high spped timer in the future.

also, I have created the pr #1850

**Reply by pkendall64** [reply]
Awesome, thanks. I've also been doing a lot of reading about how the ledc module works in the ESP32.
And re-looking at the code for the RGB LED, it actually uses a UART on esp8285 and I2S on ESP32, rather than LEDC so we should be fine to use all the ledc timers.
I'll review the PR and do some testing as well. I'll see of some of the other devs have time to do some tests as well.
